test/functional/ui/incsubstitution_spec.lua

```diff
@@ -0,0 +1,1374 @@
+    local s = ""
+    for _, buffer in pairs(t) do
+      s = s .. ", " .. tostring(buffer)
+    end
+    return s
+  end
+
+  -- fill the table with default values
+  for event, _ in pairs(eventsExpected) do
+    eventsExpected[event].open = eventsExpected[event].open or {}
+    eventsExpected[event].close = eventsExpected[event].close or {}
+  end
+
+
+  local function register_autocmd(event)
+    execute("let g:" .. event .. "_fired=[]")
+    execute("autocmd " .. event .. " * call add(g:" .. event .. "_fired, expand('<abuf>'))")
+  end
+
+  it('are run properly when splitting', function()
+    common_setup(nil, "split", default_text)
+
+    local eventsObserved = {}
+    for event, _ in pairs(eventsExpected) do
+      eventsObserved[event] = {}
+      register_autocmd(event)
+    end
+
+    feed(":%s/tw")
+
+    for event, _ in pairs(eventsExpected) do
+      eventsObserved[event].open = eval("g:" .. event .. "_fired")
+      execute("let g:" .. event .. "_fired=[]")
+    end
+
+    feed("/<enter>")
+
+    for event, _ in pairs(eventsExpected) do
+        eventsObserved[event].close = eval("g:" .. event .. "_fired") 
+    end
+
+    for event, _ in pairs(eventsExpected) do
+      eq(event .. " " .. bufferlist(eventsExpected[event].open),
+         event .. " " .. bufferlist(eventsObserved[event].open))
+      eq(event .. " " .. bufferlist(eventsExpected[event].close),
+         event .. " " .. bufferlist(eventsObserved[event].close))
+    end
+
+  end)
+
+  it('that close the incsub window', function()
+    local openEvents = {}
+
+    for event, _ in pairs(eventsExpected) do
+      if #eventsExpected[event].open > 0 then
+        openEvents[event] = true
+      end
+    end
+
+    local errorEvents = {
+      BufWipeout = "444",
+      BufDelete = "444",
+      BufUnload = "444",
+      BufWinLeave = "444",
+    }
+
+
+    local screen = Screen.new(50,10)
+
+    for event,_ in pairs(openEvents) do
+      clear()
+      common_setup(screen, "split", default_text)
+
+      execute("autocmd " .. event .. " * close!")
+      feed(":%s/.")
+
+      if errorEvents[event] then
+        feed("<enter>")
+        local err = errorEvents[event]
+        local errstr = string.find(eval('v:errmsg'), '^E' .. err .. ':')
+        neq(event .. "nil", event .. tostring(errstr))
+      else
+        screen:expect([[
+          two lines                                         |
+          {11:[No Name] [+]                                     }|
+           [1]Inc substitution on                           |
+           [2]two lines                                     |
+                                                            |
+          ~                                                 |
+          ~                                                 |
+          ~                                                 |
+          {10:[inc_sub]                                         }|
+          :%s/.^                                             |
+        ]])
```

Comment:
If I understood right then if there was no error there is still incsub window (which means that autocommands were not run I guess: if they were and there was incsub window it should’ve been closed or some other error given). But why you do not check what is there on the screen if there *was* an error? Given that it is E444 which talks about the only window left I would expect that there is no incsub window on screen.

Same applies to closing window of the original buffer.